### PR TITLE
♻️(project) move from certificate to document semantic

### PR DESCRIPTION
**Pull request by jmaupetit**
## Purpose

Marion is not restricted to certificates generation, we can consider it as generic as a documents factory.

## Proposal

- [x] move the `certificates` application to `documents`
- [x] move templates directories to `marion/documents`
- [x] `s/certificate/document/g`
- [x] `s/Certificate/Document/g`
- [x] `s/CERTIFICATE/DOCUMENT/g`

### CODE REVIEW

File: src/howard/howard/issuers/realisation.py [hunk 1]
```diff
@@ -33,8 +33,8 @@ def __str__(self):
         return f"{self.value}"
 
 
-class CertificateScope(StrEnum):
-    """Allowed scopes for the RealisationCertificate."""
+class DocumentScope(StrEnum):
+    """Allowed scopes for the RealisationDocument."""
```
Comment (jmaupetit) [inline src/howard/howard/issuers/realisation.py:37]: Ooops :sweat_smile:

File: src/howard/howard/defaults.py [hunk 2]
```diff
@@ -4,12 +4,12 @@
...
+    """Active document issuers.
 
-    Note that changing the default choices for the list of allowed certificate
+    Note that changing the default choices for the list of allowed document
     issuers may require to create and apply a database migration.
 
     """
 
-    REALISATION = "howard.issuers.RealisationCertificate", _("Realisation")
+    REALISATION = "howard.issuers.RealisationDocument", _("Realisation")
```
Comment (jmaupetit) [inline src/howard/howard/defaults.py:15]:
```suggestion
    REALISATION = "howard.issuers.RealisationCertificate", _("Realisation")
```

File: src/howard/howard/issuers/realisation.py [hunk 3]
```diff
@@ -1,4 +1,4 @@
-"""Realisation Certificate"""
+"""Realisation Document"""
```
Comment (jmaupetit) [inline src/howard/howard/issuers/realisation.py:1]:
```suggestion
"""Realisation Certificate"""
```

File: src/howard/howard/issuers/realisation.py (hunk 1 above)
Comment (jmaupetit) [reply src/howard/howard/issuers/realisation.py:37]:
```suggestion
class CertificateScope(StrEnum):
    """Allowed scopes for the RealisationCertificate."""
```

File: src/howard/howard/issuers/realisation.py [hunk 4]
```diff
@@ -116,7 +116,7 @@ class Session(BaseModel):
 
     datespan: DateSpan
     duration: int
-    scope: CertificateScope
+    scope: DocumentScope
```
Comment (jmaupetit) [inline src/howard/howard/issuers/realisation.py:119]:
```suggestion
    scope: CertificateScope
```

File: src/howard/howard/issuers/realisation.py [hunk 5]
```diff
@@ -145,7 +145,7 @@ class ContextQueryModel(BaseModel):
     course: Course
 
 
-class RealisationCertificate(AbstractCertificate):
+class RealisationDocument(AbstractDocument):
```
Comment (jmaupetit) [inline src/howard/howard/issuers/realisation.py:148]:
```suggestion
class RealisationCertificate(AbstractDocument):
```

File: src/howard/howard/issuers/realisation.py [hunk 6]
```diff
@@ -155,7 +155,7 @@ class RealisationCertificate(AbstractCertificate):
     context_query_model = ContextQueryModel
 
     def fetch_context(self, **context_query):
-        """Fetch the context that will be used to compile the certificate template."""
+        """Fetch the context that will be used to compile the document template."""
```
Comment (jmaupetit) [inline src/howard/howard/issuers/realisation.py:158]:
```suggestion
        """Fetch the context that will be used to compile the certificate template."""
```

File: src/howard/howard/tests/issuers/test_realisation.py [hunk 7]
```diff
@@ -8,15 +8,15 @@
...
     ContextModel,
     ContextQueryModel,
     Course,
     DateSpan,
+    DocumentScope,
     Gender,
     Manager,
     Organization,
-    RealisationCertificate,
+    RealisationDocument,
```
Comment (jmaupetit) [inline src/howard/howard/tests/issuers/test_realisation.py:19]:
```suggestion
    RealisationCertificate,
```

File: src/howard/howard/tests/issuers/test_realisation.py [hunk 8]
```diff
@@ -305,7 +305,7 @@ def test_student_model_with_invalid_content(kwargs, error):
                 "end": ArrowDate("2021/12/31"),
             },
             "duration": 23,
-            "scope": CertificateScope.FORMATION,
+            "scope": DocumentScope.FORMATION,
```
Comment (jmaupetit) [inline src/howard/howard/tests/issuers/test_realisation.py:308]:
```suggestion
            "scope": CertificateScope.FORMATION,
```

File: src/howard/howard/tests/issuers/test_realisation.py [hunk 9]
```diff
@@ -330,7 +330,7 @@ def test_session_model_with_valid_content(kwargs):
             {
                 "datespan": ArrowDate(date(2021, 1, 1)),
                 "duration": 23,
-                "scope": CertificateScope.FORMATION,
+                "scope": DocumentScope.FORMATION,
```
Comment (jmaupetit) [inline src/howard/howard/tests/issuers/test_realisation.py:333]:
```suggestion
                "scope": CertificateScope.FORMATION,
```

File: src/howard/howard/tests/issuers/test_realisation.py [hunk 10]
```diff
@@ -346,7 +346,7 @@ def test_session_model_with_valid_content(kwargs):
                     "end": ArrowDate("2021/12/31"),
                 },
                 "duration": "23 days",
-                "scope": CertificateScope.FORMATION,
+                "scope": DocumentScope.FORMATION,
```
Comment (jmaupetit) [inline src/howard/howard/tests/issuers/test_realisation.py:349]:
```suggestion
                "scope": CertificateScope.FORMATION,
```

File: src/howard/howard/tests/issuers/test_realisation.py [hunk 11]
```diff
@@ -378,15 +378,15 @@ def test_session_model_with_valid_content(kwargs):
                     "end": ArrowDate("2021/12/31"),
                 },
                 "duration": 23,
-                "scope": CertificateScope.FORMATION,
+                "scope": DocumentScope.FORMATION,
```
Comment (jmaupetit) [inline src/howard/howard/tests/issuers/test_realisation.py:381]:
```suggestion
                "scope": CertificateScope.FORMATION,
```

File: src/howard/howard/tests/issuers/test_realisation.py [hunk 12]
```diff
@@ -378,15 +378,15 @@ def test_session_model_with_valid_content(kwargs):
...
                 "manager": "CEO Agarwal Anant",
             },
             "manager\n  value is not a valid dict",
         ),
         (
             {
                 "duration": 23,
-                "scope": CertificateScope.FORMATION,
+                "scope": DocumentScope.FORMATION,
```
Comment (jmaupetit) [inline src/howard/howard/tests/issuers/test_realisation.py:389]:
```suggestion
                "scope": CertificateScope.FORMATION,
```

File: src/howard/howard/tests/issuers/test_realisation.py [hunk 13]
```diff
@@ -401,7 +401,7 @@ def test_session_model_with_valid_content(kwargs):
                     "start": ArrowDate("2021/01/01"),
                     "end": ArrowDate("2021/12/31"),
                 },
-                "scope": CertificateScope.FORMATION,
+                "scope": DocumentScope.FORMATION,
```
Comment (jmaupetit) [inline src/howard/howard/tests/issuers/test_realisation.py:404]:
```suggestion
                "scope": CertificateScope.FORMATION,
```

File: src/howard/howard/tests/issuers/test_realisation.py [hunk 14]
```diff
@@ -1416,14 +1416,14 @@ def test_context_query_model_with_invalid_content(kwargs, error):
         ContextQueryModel(**kwargs)
 
 
-def test_realisation_certificate_fetch_context(monkeypatch):
-    """Test Realisation Certificate fetch_context"""
+def test_realisation_document_fetch_context(monkeypatch):
```
Comment (jmaupetit) [inline src/howard/howard/tests/issuers/test_realisation.py:1419]:
```suggestion
def test_realisation_certificate_fetch_context(monkeypatch):
```

File: src/howard/howard/tests/issuers/test_realisation.py [hunk 15]
```diff
@@ -1416,14 +1416,14 @@ def test_context_query_model_with_invalid_content(kwargs, error):
...
+    """Test Realisation Document fetch_context"""
```
Comment (jmaupetit) [inline src/howard/howard/tests/issuers/test_realisation.py:1420]:
```suggestion
    """Test Realisation Certificate fetch_context"""
```

File: src/howard/howard/tests/issuers/test_realisation.py [hunk 16]
```diff
@@ -1416,14 +1416,14 @@ def test_context_query_model_with_invalid_content(kwargs, error):
...
 
-    class TestCertificate(RealisationCertificate):
+    class TestDocument(RealisationDocument):
```
Comment (jmaupetit) [inline src/howard/howard/tests/issuers/test_realisation.py:1422]:
```suggestion
    class TestCertificate(RealisationDocument):
```

File: src/howard/howard/tests/issuers/test_realisation.py [hunk 17]
```diff
@@ -1416,14 +1416,14 @@ def test_context_query_model_with_invalid_content(kwargs, error):
...
         pass
 
     identifier = uuid.uuid4()
-    test_certificate = TestCertificate(identifier=identifier)
+    test_document = TestDocument(identifier=identifier)
```
Comment (jmaupetit) [inline src/howard/howard/tests/issuers/test_realisation.py:1426]:
```suggestion
    test_certificate = TestCertificate(identifier=identifier)
```

File: src/howard/howard/tests/issuers/test_realisation.py [hunk 18]
```diff
@@ -1441,7 +1441,7 @@ class TestCertificate(RealisationCertificate):
 
     session = Session(
         datespan=datespan,
-        scope=CertificateScope.APPRENTISSAGE,
+        scope=DocumentScope.APPRENTISSAGE,
```
Comment (jmaupetit) [inline src/howard/howard/tests/issuers/test_realisation.py:1444]:
```suggestion
        scope=CertificateScope.APPRENTISSAGE,
```

File: src/howard/howard/tests/issuers/test_realisation.py [hunk 19]
```diff
@@ -1483,7 +1483,7 @@ class TestCertificate(RealisationCertificate):
         }
     )
 
-    context = test_certificate.fetch_context(
+    context = test_document.fetch_context(
```
Comment (jmaupetit) [inline src/howard/howard/tests/issuers/test_realisation.py:1486]:
```suggestion
    context = test_certificate.fetch_context(
```

File: src/marion/marion/__init__.py [hunk 20]
```diff
@@ -1,4 +1,4 @@
-"""Marion, the learning certificate factory"""
+"""Marion, the learning document factory"""
```
Comment (jmaupetit) [inline src/marion/marion/__init__.py:1]:
```suggestion
"""Marion, the documents factory"""
```

File: src/marion/marion/documents/defaults.py [hunk 21]
```diff
@@ -19,22 +19,22 @@
 CERTIFICATE_ISSUER_CHOICES_CLASS = getattr(
     settings,
     "MARION_CERTIFICATE_ISSUER_CHOICES_CLASS",
```
Comment (quitterie-lcs) [inline src/marion/marion/documents/defaults.py:21]: idem
```suggestion
    "MARION_DOCUMENT_ISSUER_CHOICES_CLASS",
```

File: src/marion/marion/documents/defaults.py [hunk 22]
```diff
@@ -19,22 +19,22 @@
...
-    "marion.certificates.defaults.CertificateIssuerChoices",
+    "marion.documents.defaults.DocumentIssuerChoices",
 )
 CERTIFICATES_ROOT = getattr(
```
Comment (quitterie-lcs) [inline src/marion/marion/documents/defaults.py:24]: idem
```suggestion
DOCUMENTS_ROOT = getattr(
```

File: src/marion/marion/documents/defaults.py [hunk 23]
```diff
@@ -19,22 +19,22 @@
...
     settings, "MARION_CERTIFICATES_ROOT", Path(settings.MEDIA_ROOT)
 )
 CERTIFICATES_TEMPLATE_ROOT = getattr(
```
Comment (quitterie-lcs) [inline src/marion/marion/documents/defaults.py:27]:
```suggestion
DOCUMENTS_TEMPLATE_ROOT = getattr(
```

File: src/marion/marion/documents/defaults.py [hunk 24]
```diff
@@ -19,22 +19,22 @@
...
-    settings, "MARION_CERTIFICATES_TEMPLATE_ROOT", Path("marion/certificates")
+    settings, "MARION_CERTIFICATES_TEMPLATE_ROOT", Path("marion/documents")
```
Comment (quitterie-lcs) [inline src/marion/marion/documents/defaults.py:28]:
```suggestion
    settings, "MARION_DOCUMENTS_TEMPLATE_ROOT", Path("marion/documents")
```

File: src/marion/marion/documents/exceptions.py [hunk 25]
```diff
@@ -0,0 +1,38 @@
+"""Exceptions for the marion.documents application"""
+
+
+class InvalidDocumentIssuer(Exception):
+    """Invalid document issuer exception.
+
+    This exception is raised when one tries to instanciate a document issuer
+    that does not exists or is not activated in the
+    CERTIFICATE_ISSUER_CHOICES_CLASS setting (see defaults).
```
Comment (quitterie-lcs) [inline src/marion/marion/documents/exceptions.py:9]:
```suggestion
    DOCUMENT_ISSUER_CHOICES_CLASS setting (see defaults).
```

File: src/marion/marion/documents/issuers/base.py [hunk 26]
```diff
@@ -208,27 +208,25 @@ def generate_identifier(self, identifier=None):
...
-        Return default (or set) certificate path as a pathlib.Path object.
+        Return default (or set) document path as a pathlib.Path object.
 
         """
 
-        if hasattr(self, "certificate_path") and self.certificate_path is not None:
-            return self.certificate_path
+        if hasattr(self, "document_path") and self.document_path is not None:
+            return self.document_path
         return defaults.CERTIFICATES_ROOT.joinpath(f"{self.identifier}.pdf")
```
Comment (quitterie-lcs) [inline src/marion/marion/documents/issuers/base.py:220]:
```suggestion
        return defaults.DOCUMENTS_ROOT.joinpath(f"{self.identifier}.pdf")
```

File: src/marion/marion/documents/issuers/base.py [hunk 27]
```diff
@@ -208,27 +208,25 @@ def generate_identifier(self, identifier=None):
...
+        """Get (generated) document URL.
 
         If the host argument is provided a fully qualified URL will be
         returned, or else, an absolute URL will be generated.
 
         """
-        relative_path = self.get_certificate_path().relative_to(
-            defaults.CERTIFICATES_ROOT
-        )
+        relative_path = self.get_document_path().relative_to(defaults.CERTIFICATES_ROOT)
```
Comment (quitterie-lcs) [inline src/marion/marion/documents/issuers/base.py:229]:
```suggestion
        relative_path = self.get_document_path().relative_to(defaults.DOCUMENTS_ROOT)
```

File: src/marion/marion/documents/models.py [hunk 28]
```diff
@@ -11,9 +11,9 @@
 from pydantic.error_wrappers import ValidationError as PydanticValidationError
 
 from .defaults import CERTIFICATE_ISSUER_CHOICES_CLASS
```
Comment (quitterie-lcs) [inline src/marion/marion/documents/models.py:13]:
```suggestion
from .defaults import DOCUMENT_ISSUER_CHOICES_CLASS
```

File: src/marion/marion/documents/models.py [hunk 29]
```diff
@@ -11,9 +11,9 @@
...
-from .exceptions import InvalidCertificateIssuer
+from .exceptions import InvalidDocumentIssuer
 
-CertificateIssuerChoices = import_string(CERTIFICATE_ISSUER_CHOICES_CLASS)
+DocumentIssuerChoices = import_string(CERTIFICATE_ISSUER_CHOICES_CLASS)
```
Comment (quitterie-lcs) [inline src/marion/marion/documents/models.py:16]:
```suggestion
DocumentIssuerChoices = import_string(DOCUMENT_ISSUER_CHOICES_CLASS)
```

File: src/marion/marion/documents/tests/test_defaults.py [hunk 30]
```diff
@@ -0,0 +1,34 @@
...
+import importlib
+from pathlib import Path
+
+from marion.documents import defaults
+
+
+def test_defaults_overrides_with_settings(settings):
+    """Test marion.documents.defaults overrides with settings definition"""
+
+    settings.MARION_CERTIFICATE_ISSUER_CHOICES_CLASS = (
```
Comment (quitterie-lcs) [inline src/marion/marion/documents/tests/test_defaults.py:12]:
```suggestion
    settings.MARION_DOCUMENT_ISSUER_CHOICES_CLASS = (
```

File: src/marion/marion/documents/tests/test_defaults.py [hunk 31]
```diff
@@ -0,0 +1,34 @@
...
+        "howard.documents.DocumentIssuerChoices"
+    )
+    settings.MARION_CERTIFICATES_ROOT = Path("/tmp/documents/abc")
```
Comment (quitterie-lcs) [inline src/marion/marion/documents/tests/test_defaults.py:15]:
```suggestion
    settings.MARION_DOCUMENTS_ROOT = Path("/tmp/documents/abc")
```

File: src/marion/marion/documents/tests/test_defaults.py [hunk 32]
```diff
@@ -0,0 +1,34 @@
...
+    settings.MARION_CERTIFICATES_TEMPLATE_ROOT = Path("howard/documents/abc")
```
Comment (quitterie-lcs) [inline src/marion/marion/documents/tests/test_defaults.py:16]:
```suggestion
    settings.MARION_DOCUMENTS_TEMPLATE_ROOT = Path("howard/documents/abc")
```

File: src/marion/marion/documents/tests/test_defaults.py [hunk 33]
```diff
@@ -0,0 +1,34 @@
...
+
+    # Force module reload to take into account setting override as it is loaded
+    # very early in the stack
+    importlib.reload(defaults)
+
+    assert (
+        defaults.CERTIFICATE_ISSUER_CHOICES_CLASS
```
Comment (quitterie-lcs) [inline src/marion/marion/documents/tests/test_defaults.py:23]:
```suggestion
        defaults.DOCUMENT_ISSUER_CHOICES_CLASS
```

File: src/marion/marion/documents/tests/test_defaults.py [hunk 34]
```diff
@@ -0,0 +1,34 @@
...
+        == "howard.documents.DocumentIssuerChoices"
+    )
+    assert defaults.CERTIFICATES_ROOT == Path("/tmp/documents/abc")
```
Comment (quitterie-lcs) [inline src/marion/marion/documents/tests/test_defaults.py:26]:
```suggestion
    assert defaults.DOCUMENTS_ROOT == Path("/tmp/documents/abc")
```

File: src/marion/marion/documents/tests/test_defaults.py [hunk 35]
```diff
@@ -0,0 +1,34 @@
...
+    assert defaults.CERTIFICATES_TEMPLATE_ROOT == Path("howard/documents/abc")
```
Comment (quitterie-lcs) [inline src/marion/marion/documents/tests/test_defaults.py:27]:
```suggestion
    assert defaults.DOCUMENTS_TEMPLATE_ROOT == Path("howard/documents/abc")
```

File: src/marion/marion/documents/tests/test_views.py [hunk 36]
```diff
@@ -13,27 +13,27 @@
...
     return len(list(root.glob("*.pdf")))
 
 
 @pytest.mark.django_db
-def test_certificate_request_viewset_post(monkeypatch):
-    """Test the CertificateRequestViewSet create view"""
+def test_document_request_viewset_post(monkeypatch):
+    """Test the DocumentRequestViewSet create view"""
 
     monkeypatch.setattr(defaults, "CERTIFICATES_ROOT", Path(tempfile.mkdtemp()))
```
Comment (quitterie-lcs) [inline src/marion/marion/documents/tests/test_views.py:32]:
```suggestion
    monkeypatch.setattr(defaults, "DOCUMENTS_ROOT", Path(tempfile.mkdtemp()))
```

File: src/marion/marion/documents/tests/test_views.py [hunk 37]
```diff
@@ -13,27 +13,27 @@
...
 
-    url = reverse("certificaterequest-list")
+    url = reverse("documentrequest-list")
 
-    assert count_certificates(defaults.CERTIFICATES_ROOT) == 0
+    assert count_documents(defaults.CERTIFICATES_ROOT) == 0
```
Comment (quitterie-lcs) [inline src/marion/marion/documents/tests/test_views.py:36]:
```suggestion
    assert count_documents(defaults.DOCUMENTS_ROOT) == 0
```

File: src/marion/marion/documents/tests/test_views.py [hunk 38]
```diff
@@ -43,77 +43,77 @@ def test_certificate_request_viewset_post(monkeypatch):
     assert response.data.get("context_query")[0].code == "required"
     assert isinstance(response.data.get("issuer")[0], drf_exceptions.ErrorDetail)
     assert response.data.get("issuer")[0].code == "required"
-    assert models.CertificateRequest.objects.count() == 0
-    assert count_certificates(defaults.CERTIFICATES_ROOT) == 0
+    assert models.DocumentRequest.objects.count() == 0
+    assert count_documents(defaults.CERTIFICATES_ROOT) == 0
```
Comment (quitterie-lcs) [inline src/marion/marion/documents/tests/test_views.py:47]:
```suggestion
    assert count_documents(defaults.DOCUMENTS_ROOT) == 0
```

File: src/marion/marion/documents/tests/test_views.py [hunk 39]
```diff
@@ -43,77 +43,77 @@ def test_certificate_request_viewset_post(monkeypatch):
...
+        "issuer": "marion.documents.issuers.DumberDocument",
         "context_query": json.dumps({"fullname": "Richie Cunningham"}),
     }
     response = client.post(url, data, format="json")
     assert response.status_code == status.HTTP_400_BAD_REQUEST
     assert response.data.get("issuer")[0].code == "invalid_choice"
-    assert models.CertificateRequest.objects.count() == 0
-    assert count_certificates(defaults.CERTIFICATES_ROOT) == 0
+    assert models.DocumentRequest.objects.count() == 0
+    assert count_documents(defaults.CERTIFICATES_ROOT) == 0
```
Comment (quitterie-lcs) [inline src/marion/marion/documents/tests/test_views.py:58]:
```suggestion
    assert count_documents(defaults.DOCUMENTS_ROOT) == 0
```

**Review by quitterie-lcs — COMMENTED** [review]
(At the end, I did not comment the settings still with CERTIFICATES semantic in test)

**Comment by jmaupetit** [comment]
@quitterie-lcs This is all fixed now!

**Review by quitterie-lcs — APPROVED** [review]

**Pull request merged**